### ISSUE TITLE
How is the `logging.InjectLogField` supposed to work?

### ISSUE THREAD

**Issue by LasseJacobs** [grpc-ecosystem/go-grpc-middleware#597]
The logging interceptor supports inject custom fields via the `InjectLogField` function. This function will 'inject' the key-value pair into the context by creating a copy of the existing list of fields, appending the new field and setting the new list to the context key. This keeps the values in the context immutable. However that also means you cannot add any fields to the logger inside any middleware of RPC that comes after the logging interceptor.

Take the example from the repo, if you add an interceptor after the `logging.UnaryServerInterceptor` that will add fields via `InjectLogField` they will not have any effect:
```
// You can now create a server with logging instrumentation that e.g. logs when the unary or stream call is started or finished.
	_ = grpc.NewServer(
		grpc.ChainUnaryInterceptor(
			logging.UnaryServerInterceptor(InterceptorLogger(logger), opts...),
			// Add any other interceptor you want.
                        SomeUnaryServerInterceptorThatTriesToAddFields(),    // <- this will not work
		),
		grpc.ChainStreamInterceptor(
			logging.StreamServerInterceptor(InterceptorLogger(logger), opts...),
			// Add any other interceptor you want.
		),
	)
	// ...user server.
```

If you place the interceptor to come before the logger it works if you set the fields _before_ calling the handler.
```
grpc.ChainUnaryInterceptor(
        SomeUnaryServerInterceptorThatTriesToAddFields(),    //  this works if InjectLogField is called before calling the handler
	logging.UnaryServerInterceptor(InterceptorLogger(logger), opts...),
	// Add any other interceptor you want.
),
```

But that means you can never set any fields based on the response of the RPC call. Is this how this method is supposed to work? What I would like is if the interceptor could do:
```
func UnaryServerInterceptor() grpc.UnaryServerInterceptor {
	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
		resp, err := handler(ctx, req)
		if err != nil {
			// extract some custom fields from the error
			...
			// set those fields to the log:
			grpc_logging.InjectLogField(ctx, "custom field", customfields)
		}
		return resp, err
	}
}
```

But this obviously does not work, there is not even a way to update the context. Any suggested work around? I can store a pointer into the context with one interceptor that comes before the logging interceptor (so it can pass the context), and set a value with an interceptor that comes after the logging interceptor  in the chain (so that it will run before the logging interceptor). But I am hoping for a more elegant way?

**Comment by ti** [comment]
Same problem，`logging.InjectLogField` does not work with zap logger

**Comment by johanbrandhorst** [comment]
I am not sure but I think this is generally working as intended - you add things to the context so that "inner" functions can access them. If you want to add something to the response, you could use grpc.SetHeader or something like that maybe?

**Comment by LasseJacobs** [comment]
I don't want to add anything to the response, but I want to add fields to the logger based on the response. I tried a bunch of things but in the end contexts are meant to be immutable so you really can only pass things down.

The work around I used is to store a pointer in the context to a map with fields. Since it is a pointer I can update the maps 'downstream' and changes will be visible in the context that is passed to the logging interceptor.

I don't think the problem has to do with the type of logger used, it is just the way context works, it is 'immutable' and the fields are stored as values so only changes made by interceptors that come before the logging interceptor will take effect.
